aequilibrae/project/network/connector_creation.py

```diff
@@ -0,0 +1,124 @@
+
+    # We expand the area by its average radius until it is 20 times
+    # beginning with a strict search within the zone
+    buffer = 0
+    increase = sqrt(geo.area / pi)
+    dt = []
+    while dt == [] and buffer <= increase * 10:
+        wkb = geo.buffer(buffer).wkb
+        curr.execute(sql, [wkb, srid, wkb, srid, lt, mode_id])
+        dt = curr.fetchall()
+        buffer += increase
+
+    if buffer > increase:
+        msg = f'Could not find node inside zone {zone_id}. Search area was expanded until we found a suitable node'
+        logger.warning(msg)
+    if dt == []:
+        warn(f'FAILED! Could not find suitable nodes to connect within 5 times the diameter of zone {zone_id}.')
+        return
+
+    coords = []
+    nodes = []
+    for node_id, wkb, modes, link_types in dt:
+        geo = shapely.wkb.loads(wkb)
+        coords.append([geo.x, geo.y])
+        nodes.append(node_id)
+
+    num_connectors = connectors
+    if len(nodes) == 0:
+        raise Exception('We could not find any candidate nodes that satisfied your criteria')
+    elif len(nodes) < connectors:
+        warn(f'We have fewer possible nodes than required connectors for zone {zone_id}. Will connect all of them.')
+        num_connectors = len(nodes)
+
+    if num_connectors == len(coords):
+        all_nodes = nodes
+    else:
+        features = np.array(coords)
+        whitened = whiten(features)
+        centroids, allocation = kmeans2(whitened, num_connectors)
+
+        all_nodes = set()
+        for i in range(num_connectors):
+            nds = [x for x, y in zip(nodes, list(allocation)) if y == i]
+            centr = centroids[i]
+            positions = [x for x, y in zip(whitened, allocation) if y == i]
+            if positions:
+                dist = cdist(np.array([centr]), np.array(positions)).flatten()
+                node_to_connect = nds[dist.argmin()]
+                all_nodes.add(node_to_connect)
+
+    nds = list(all_nodes)
+    data = [zone_id] + nds
+    curr.execute(f'select b_node from links where a_node=? and b_node in ({",".join(["?"] * len(nds))})', data)
+
+    data = [x[0] for x in curr.fetchall()]
+
+    if data:
+        qry = ",".join(["?"] * len(data))
+        dt = [mode_id, zone_id] + data
+        curr.execute(f'Update links set modes=modes || ? where a_node=? and b_node in ({qry})', dt)
+        nds = [x for x in nds if x not in data]
+        logger.warning(f'Mode {mode_id} added to {len(data)} existing centroid connectors for zone {zone_id}')
+        conn.commit()
+
+    curr.close()
+    links = network.Links()
+
+    for node_to_connect in nds:
+        link = links.new()
+        node_to = proj_nodes.get(node_to_connect)
+        link.geometry = LineString([node.geometry, node_to.geometry])
+        link.modes = mode_id
+        link.direction = 0
+        link.link_type = 'centroid_connector'
+        link.name = f'centroid connector zone {zone_id}'
+        link.capacity_ab = 99999
```

Comment:
Should we keep the default capacity as a constant in one place for clarity?

Reply:
Addressed